File: logging_middleware/logger.py
import os
import json
import urllib.request
import urllib.error

LOG_URL = "http://20.207.122.201/evaluation-service/logs"

VALID_LEVELS = {"debug", "info", "warn", "error", "fatal"}
VALID_STACKS = {"backend", "frontend"}

BACKEND_PACKAGES = {
    "cache", "controller", "cron_job", "db", "domain",
    "handler", "repository", "route", "service"
}
FRONTEND_PACKAGES = {"api", "component", "hook", "page", "state", "style"}
SHARED_PACKAGES   = {"auth", "config", "middleware", "utils"}


def _is_valid_package(stack: str, package: str) -> bool:
    if package in SHARED_PACKAGES:
        return True
    if stack == "backend":
        return package in BACKEND_PACKAGES
    if stack == "frontend":
        return package in FRONTEND_PACKAGES
    return False


def log(stack: str, level: str, package: str, message: str) -> None:
    if stack not in VALID_STACKS:
        print(f"[logger] skipped — invalid stack: '{stack}'")
        return

    if level not in VALID_LEVELS:
        print(f"[logger] skipped — invalid level: '{level}'")
        return

    if not _is_valid_package(stack, package):
        print(f"[logger] skipped — '{package}' not allowed for stack '{stack}'")
        return

    token = os.environ.get("AUTH_TOKEN", "")
    if not token:
        print("[logger] AUTH_TOKEN not set — skipping remote log")
        return

    payload = json.dumps({
        "stack":   stack,
        "level":   level,
        "package": package,
        "message": message,
    }).encode("utf-8")

    req = urllib.request.Request(
        url=LOG_URL,
        data=payload,
        method="POST",
        headers={
            "Content-Type":  "application/json",
            "Authorization": f"Bearer {token}",
        },
    )

    try:
        with urllib.request.urlopen(req, timeout=5) as resp:
            result = json.loads(resp.read().decode())
            if "logID" in result:
                print(f"[logger] sent ({level}) → logID: {result['logID']}")
    except urllib.error.URLError as e:
        print(f"[logger] request failed: {e.reason}")
    except Exception as e:
        print(f"[logger] unexpected error: {e}")

File: logging_middleware/logger.js
const axios = require('axios');
const fs = require('fs');
const path = require('path');

const LOG_URL = "http://20.207.122.201/evaluation-service/logs";

const VALID_LEVELS = new Set(["debug", "info", "warn", "error", "fatal"]);
const VALID_STACKS = new Set(["backend", "frontend"]);

const BACKEND_PACKAGES = new Set([
  "cache", "controller", "cron_job", "db", "domain",
  "handler", "repository", "route", "service"
]);
const SHARED_PACKAGES = new Set(["auth", "config", "middleware", "utils"]);

function isValidPackage(stack, pkg) {
  if (SHARED_PACKAGES.has(pkg)) return true;
  if (stack === "backend") return BACKEND_PACKAGES.has(pkg);
  return false;
}

async function Log(stack, level, pkg, message) {
  if (!VALID_STACKS.has(stack)) {
    console.log(`[logger] skipped — invalid stack: '${stack}'`);
    return;
  }
  if (!VALID_LEVELS.has(level)) {
    console.log(`[logger] skipped — invalid level: '${level}'`);
    return;
  }
  if (!isValidPackage(stack, pkg)) {
    console.log(`[logger] skipped — '${pkg}' not allowed for stack '${stack}'`);
    return;
  }

  const envPath = path.join(__dirname, '..', 'notification_app_be', '.env');
  let token = process.env.AUTH_TOKEN;
  
  if (!token && fs.existsSync(envPath)) {
    const envContent = fs.readFileSync(envPath, 'utf8');
    const match = envContent.match(/AUTH_TOKEN=(.*)/);
    if (match) token = match[1].trim();
  }

  if (!token) {
    console.log("[logger] AUTH_TOKEN not found — skipping remote log");
    return;
  }

  try {
    const response = await axios.post(LOG_URL, {
      stack,
      level,
      package: pkg,
      message
    }, {
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      }
    });

    if (response.data && response.data.logID) {
      console.log(`[logger] sent (${level}) → logID: ${response.data.logID}`);
    }
  } catch (err) {
    console.log(`[logger] request failed: ${err.message}`);
  }
}

module.exports = { Log };
